### DISCUSSION TITLE
Firestore sync backend being implemented

### DISCUSSION THREAD

**Discussion by anicolao** [TiddlyWiki/TiddlyWiki5#7773] [General]
I looked at the existing firestore plugin and it wasn't at all what I wanted, so I wrote what I wanted.

This has resulted in a very early proof of concept and I would welcome feedback here on where to take it / what features people would like to see. The project does not involve modifying TiddlyWiki5, but is built to work with the existing system.

There are two parts to the project:

1. A firestore sync plugin, that uses the existing sync mechanism to save all your tiddlers to a firestore database bi-directionally. The effect of this on a vanilla tw5 empty instance is that all viewers of that instance see the same state: the same story river, the same tiddlers, the same drafts. They can even attempt to "collaboratively" edit a draft, though the delays implemented in the core for how this works make the experience a little janky. Nonetheless it is a "multi-user" wiki where all users see and edit the same state simultaneously.

2. A meta website which lets you click a button to generate a new URL configured to point at a firestore backend _that you own_. This works by asking you permission to modify your google cloud projects, creating a project named tw5-yourwikiname-randomdigits, creating a firestore database in it for you, configuring security rules on your backend database, and then serving your URL so that when you click the link to open your wiki you are now looking at your content on your backend.

From here, many different directions could be pursued. The wikis could be sharable, with multi-user collaboration and editing; the wikis could be entirely private; it might be nice to have collaborative editing working smoothly in the wikis; etc, etc.

I am very keen for feedback/input if this sounds interesting to others.

**Comment by pmario** [comment]
That's interesting, but without seeing anything it's hard to give feedback. 

What's about the pricing?
What if I'm DDOSed from a region, which is expensive?

**Reply by anicolao** [reply]
The cloud project that is provisioned does not have billing enabled, so it is not possible to have to pay without enabling it manually. The free tier permits 50,000 reads per day and 20,000 writes per day, which I believe is sufficient for most people's personal wiki notes requirements; but if that is not enough, you can enable billing on the storage at a cost of $0.06 USD per 100,000 reads/40k writes. 

Currently TiddlyWiki seems to read a little more than necessary, and I will look into reducing the read count; but it also could be because I left 50 tabs open and so was reading the wiki 50x every time I edited it (all viewers update in real time as the wiki is updated).

For DOS or DDOS attacks, the attacker would need to know your secrets in order to attack the service. It is not openly accessible to someone who doesn't have the config. However I do want to enable sharing of the wiki, so you will be at the mercy of those you share it with. If you share globally, you'd need to think about DDOS but for now I am not worrying about it as it is not really possible for anyone to hit your wiki; it is essentially private storage for you.

**Reply by anicolao** [reply]
I will publish a link once it is a little more polished. But one challenge is that the current setup needs an OAuth scope to set up your cloud project for you; you can also set it up manually but that is well beyond most users' capabilities. So figuring out how to make that less scary is a problem.

**Reply by pmario** [reply]
Thanks for the info. -- Don't get me wrong, I do like the project and I'm looking forward to the demo but I have to play "The Devil's Advocate" 

We do need to do some math, because most users do not. 

I do not know how your sync mechanism exactly works but you wrote it reads real-time. At the moment TW stores every change to a tiddler in edit-mode, 400ms after typing stopped, into a "Draft of xxx" tiddler. 

Since most users make longer pauses than that after a view characters, we can say that TW syncs back to the server every second if work is done. 

With that in mind 20000 writes are not as much as it seems. An hour has 3600 seconds. So after 5 1/2 hours of intense work, we will reach the limit. 

In your OP you wrote, that the full TW state is synced back to the server. That means every click on the UI will be synced too. So the limit will be reached much earlier. Every UI interaction can change several tiddlers at once. 

For the node-js server we avoid to sync back every tiddler that is prefixed `$:/state/`, `$:/temp/` [and several more](https://tiddlywiki.com/#%24%3A%2Fconfig%2FSyncFilter) for exactly that reason. 

The $:/StoryList is synced back to the server, but not promoted to connected clients. Since everyone needs to see the same thing at the same time. This may be nice for a presentation, but for multi-user mode it will not really work. 

The TW [default syncer](https://tiddlywiki.com/#%24%3A%2Fcore%2Fmodules%2Fsyncer.js) polls the whole tiddler store every minute. 1 day has 14400 minutes. So if 7 tabs in "multi-user mode" are active in browsers the read limit of 50000 reads will be reached **by doing nothing** after about 12 hours. 

So idle users can be heavy on the budget.

From experience and feedback at the discussion groups I know that we have many users that do not close there browser tabs for months. TW has [some hidden settings](https://tiddlywiki.com/#Hidden%20Setting%3A%20Sync%20System%20Tiddlers%20From%20Server:%5B%5BHidden%20Setting%3A%20Sync%20System%20Tiddlers%20From%20Server%5D%5D%20%5B%5BHidden%20Setting%3A%20Sync%20Polling%20Interval%5D%5D), that you may not know, that let us adjust some of those parameters

**Reply by pmario** [reply]
I also think the "fee tier" is per account, so it has to be shared by all wikis. -- right?

**Reply by anicolao** [reply]
These details are very helpful, because they'll enable us to work out what needs to be modified to make it work, thanks!

From a TW5 perspective, my plugin uses the existing sync mechanism, with all of its current design decisions. For example, a common bug in the current sync mechanism is that it fails to do the last write, so the very last thing the user did is never written to the store, usually just one character or a fragment of a word at my typing speed. System tiddler syncing is off by default, but I wish to turn it on because without it you cannot install additional plugins in a meaningful way.

When you say the default syncer polls the whole store every minute, what does that mean exactly? I have not seen this behaviour in my plugin, but I may not be paying enough attention. Does it ask the plugin to read every document every minute? It's this kind of usage - not tied to user behaviour - that will ultimately be most problematic for the free tier. I completely agree that users will just leave tabs open, potentially multiple tabs on the same computer, 24/7/365, and that can become a problem.

The free tier is per Google Cloud project. When the user uses the meta project to create their wiki, it makes a new cloud project, so each wiki is in a new free tier. The user can run out of Google Cloud projects relatively easily if they create dozens of wikis, but until they do each one gets a new free tier to play in, and each one is separate (that is one could go quota exceeded while the others work fine). 

Sorry for the delay posting a link; I will get a link to the deployed meta project up once the UI has a little feedback on what's going on. Provisioning the cloud project takes time and in testing it with new accounts I found that if the user has never accepted Firebase's terms of service it doesn't work. I need to fix these problems before putting up a link you can try.

**Reply by pmario** [reply]
>Does it ask the plugin to read every document every minute?

There is a GET command that is automatically triggered by the client every minute.  See the screen-shot

**F12 - Network tab**

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/368161c1-7a4e-4f35-8a58-5944eb987bd8)

**Request - Skinny tiddlers**

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/76746b06-d2a5-45e5-9c01-e8a4b263ac14)

**Response**

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/b534c746-25b4-4390-bd45-8ecee67ac0dd)

**Reply by pmario** [reply]
> The free tier is per Google Cloud project. 

That's good to know

**Reply by pmario** [reply]
>Sorry for the delay posting a link;

No problem. Take your time

**Comment by anicolao** [comment]
In the edition I am running, which is `empty` plus my sync plugin, this request doesn't happen. Coincidentally there is a bit of traffic to firebase every minute, but it is empty if the user is inactive and doesn't result in any billable read. A billable read from firebase has to actually read a document, which means the underlying document has to have been changed by the user. So I am not quite sure how to repro the get requests you are showing, but at least for now it doesn't seem to be a concern. Here's a screenshot after about 7 minutes:

<img width="1440" alt="image" src="https://github.com/Jermolene/TiddlyWiki5/assets/1145048/7ca08c9f-e689-4469-932a-eecdca730952">

Here is a URL at which you can try it out: https://tiddlywiki-a94cd.web.app/

If you're concerned about the OAuth token being obtained, which allows my code to manipulate your google cloud state for up to 1h after you sign in, make a new Google account to try it out with. If you make a new google account (or if you've never used firebase before) a defect in the current implementation is that it can't accept the terms of service for you or take you through a ToS flow, so you have to go create a throwaway firebase project on the usual console first, and then come back and click the create wiki button. Additionally, if you use a throwaway account you can be sure Google cannot bill you no matter what bugs might be present in my implementation or in TW5 itself, so it's a nice way to have a playground to look at the project in.

Once you've created your wiki, if you share it iwth me (anicolao) using my gmail.com as my email address, I will be able to see it and collaborate on your tiddlywiki with you.

**Comment by linonetwo** [comment]
Are you talking about syncer or sync-adaptor? With sync-adaptor you won't need to 

> save all your tiddlers to a firestore

You can just save what is changed. See https://github.com/Jermolene/TiddlyWiki5/discussions/7720

**Reply by anicolao** [reply]
The plugin is a sync adaptor. It doesn't save redundantly. But the intention is to save everything (once) so that your entire wiki is in your firestore storage, and when you reload it from a new device you're exactly where you left off.

**Reply by linonetwo** [reply]
Cool, So you store all tiddlers as rows in database, and when open webpage again later, you can lazy-load then?

I was trying to do this with SoLiD in https://github.com/linonetwo/solid-tiddlywiki-syncadaptor , but I was not so familiar with tiddlywiki at that time. I can learn from your code to upgrade the solid syncadaptor.
I'm also doing this in https://github.com/tiddly-gittly/TidGi-Mobile/blob/44842252196093ebe49ebe1397f08436f364bd9a/plugins/src/expo-file-system-syncadaptor/file-system-syncadaptor.ts to read tiddlers from local sqlite db.

**Reply by anicolao** [reply]
My (very early version) sync plugin code is here: https://github.com/anicolao/tw5-firestore-sync

It doesn't lazy load; on startup it will read all tiddlers into memory. After that, it reads and writes only changed tiddlers. With firebase it is easy to enable local caching (which I have not done yet) so that the startup load comes from indexeddb on the client instead of over the network, so this should be fast enough even for largish wikis.

**Comment by linonetwo** [comment]
![图片](https://github.com/Jermolene/TiddlyWiki5/assets/3746270/8a41119e-1028-4425-b42f-3a1de0659de4)

Maybe it is hard for me to access firebase in my region? It just stop here.

**Reply by anicolao** [reply]
Sorry I need to improve the error handling. This is because you've -never- created a firebase project before. As the text at the top of the pge says, if you've never used firebase you have to go to the firebase console and create any project to agree to the terms of service. Then you can delete that project or just leave it there, and come back - the 'adding firebase' will then succeed. Until you've aaccepted the terms of service, it gets stuck here. Please do try again!

**Reply by anicolao** [reply]
In the [firebase console](https://console.firebase.google.com/), once you reach a page like the one shown below for your new project, coming back and trying again should work.
<img width="598" alt="image" src="https://github.com/Jermolene/TiddlyWiki5/assets/1145048/1cdaf063-b8cf-49b6-b017-66354756f1db">
